contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,652 @@
+      mstore(add(fp, 0x40), mload(add(add(proof, blockNumOffset), 0x60))) // numWords
+      sender := mload(add(add(proof, blockNumOffset), 0x80))
+    }
+    currentKeyHash = hashOfKey(publicKey);
+    bytes32 callback = s_callbacks[preSeed];
+    requestId = preSeed;
+    if (callback == 0) {
+      revert NoCorrespondingRequest();
+    }
+    if (callback != keccak256(abi.encodePacked(
+        requestId,
+        blockNum,
+        fp.subId,
+        fp.callbackGasLimit,
+        fp.numWords,
+        sender)))
+    {
+      revert IncorrectCommitment();
+    }
+    fp.sender = sender;
+
+    bytes32 blockHash = blockhash(blockNum);
+    if (blockHash == bytes32(0)) {
+      blockHash = BLOCKHASH_STORE.getBlockhash(blockNum);
+      if (blockHash == bytes32(0)) {
+        revert BlockhashNotInStore(blockNum);
+      }
+    }
+    // The seed actually used by the VRF machinery, mixing in the blockhash
+    uint256 actualSeed = uint256(keccak256(abi.encodePacked(preSeed, blockHash)));
+    // solhint-disable-next-line no-inline-assembly
+    assembly { // Construct the actual proof from the remains of proof
+      mstore(add(proof, PRESEED_OFFSET), actualSeed)
+      mstore(proof, PROOF_LENGTH)
+    }
+    randomness = VRF.randomValueFromVRFProof(proof); // Reverts on failure
+  }
+
+  function getFeedData()
+    private
+    view
+    returns (
+      int256
+    )
+  {
+    uint32 stalenessSeconds = s_config.stalenessSeconds;
+    bool staleFallback = stalenessSeconds > 0;
+    uint256 timestamp;
+    int256 weiPerUnitLink;
+    (,weiPerUnitLink,,timestamp,) = LINK_ETH_FEED.latestRoundData();
+    if (staleFallback && stalenessSeconds < block.timestamp - timestamp) {
+      weiPerUnitLink = s_fallbackWeiPerUnitLink;
+    }
+    return weiPerUnitLink;
+  }
+
+  function withdraw(
+    address recipient,
+    uint96 amount
+  )
+    external
+  {
+    if (s_withdrawableTokens[msg.sender] < amount) {
+      revert InsufficientBalance();
+    }
+    s_withdrawableTokens[msg.sender] -= amount;
+    if (!LINK.transfer(recipient, amount)) {
+      revert InsufficientBalance();
+    }
+  }
+
+  function getSubscription(
+    uint64 subId
+  )
+    external
+    view
+    returns (
+      uint96 balance,
+      address owner,
+      address[] memory consumers
+    )
+  {
+    return (
+      s_subscriptions[subId].balance,
+      s_subscriptions[subId].owner,
+      s_subscriptions[subId].consumers
+    );
+  }
+
+  function createSubscription(
+    address[] memory consumers // permitted consumers of the subscription
+  )
+    external
+    returns (
+      uint64
+    )
+  {
+    currentSubId++;
+    s_subscriptions[currentSubId] = Subscription({
+      balance: 0,
+      owner: msg.sender,
+      requestedOwner: address(0),
+      consumers: consumers
+    });
+    emit SubscriptionCreated(currentSubId, msg.sender, consumers);
+    return currentSubId;
+  }
+
+  function requestSubscriptionOwnerTransfer(
```

Comment:
Could we add some docblocks for these external functions, please?

Reply:
I was holding off on that until logic review is complete and the API is stable. Planning to do it in a follow up PR